### DISCUSSION TITLE
[Routing / Navigation question] How to load / handle query params?

### DISCUSSION THREAD

**Discussion by dessalines** [framesurge/perseus#277] [Q&A] [answered]
I'm making a simple HTTP API fetching front end, that takes something like `/search?q=...` . 

Are there any examples / demos of Query params with perseus?

**Comment by arctic-hen7** [comment]
Is `/search` going to be a page on your site? If so, what you want is a `search` template that creates a page for each search, which is probably best done with incremental generation (reading through the docs on the template/page system in Perseus may help you to understand this). 

If you want to have *one* page using request-time state to return details for a search in an actual URL query parameter, then you would need access to the query parameters through the `Request<()>` object that gets passed through. That's from the `http` crate, so maybe check how to access query parameters through there? If that doesn't work, then I'll convert this into an issue asking for that feature. Until that's implemented, maybe use `#q=...` instead, and have the client perform an API request after a generic page loads to fetch search-specific data.

**Reply by dessalines** [reply]
Yes, to outline what I'd like to do (and these are all very common use-cases)

- Have a `search` template, containing just a search box, but needing no pre-fetched state.
- Do a search, browser fetch from an API, have that update the page query params: `search?q=MY_SEARCH_TERM` . Whether the search does any real navigation or faux-navigation is immaterial.
- If a user goes directly to the page `search?q=MY_SEARCH_TERM`, then the server-side should pre-fetch that data.

I see that sycamore has *some* routing / navigation docs pertaining to variable routes IE `/search/<term>`, but I didn't see anything in perseus.

> Until that's implemented, maybe use #q=... instead, and have the client perform an API request after a generic page loads to fetch search-specific data.

I think that's going to have to be the workaround for now. But yes I'd like to stress that a combined BrowserRouter and StaticRouter is ubiquitous for any app that has variable data : `product?id={}`, `post/{id}?sort=new&filter=....`, etc.

**Reply by arctic-hen7** [reply]
I think you should read up on the Perseus templating system in the docs, because this is all baked in at that level. Routing in Perseus is implicit, meaning you don't have to specify "do this when the user goes here", Perseus will figure it out. The ergonomic thing to do here is to use incremental generation and have one template for your search box page and another for your search results pages. However, assuming you don't want to cache search results for future users (which incremental generation would let you do), I would advise using the second approach I described.

That said, I'm willing to look at a feature to allow incremental generation with request state alone, so that searches don't need to be cached at all...

**Reply by dessalines** [reply]
I can definitely have a `/search` page, and a `/results` page (once the async search is completed on the client), but there's no declarative / variable path routing with any of that.

I couldn't send someone a link with `search/q=a_query`, or `/product/my_product_id` and have the server do the fetching on the back end. Or at least I don't see any to do that in the docs or examples.

**Reply by dessalines** [reply]
I've finished creating a sample [perseus app btw (uses routing, and server-side http fetching)](https://git.torrents-csv.ml/heretic/torrents-csv-ui-perseus)! I'm really enjoying working with it. Thank you so much for making!

I'd like to use it for larger projects, but I *really* would like some more docs / examples around routing / `.incremental_generation()`

Namely, how would you handle something like: `/search?q={term}?sort={sort}` with server side fetching? Currently I'm doing `/search/term`, and I'm not sure how I would add other params.

**Reply by arctic-hen7** [reply, marked as answer]
Incremental generation handles all that for you! Your build state function will take in a path, and you can so anything you want from there: all the logic for fetching is in there based on the path you get. From there, routing is implicit. 

The `state_generation` example should cover most of this?

**Reply by dessalines** [reply]
Ah I didn't see that example, it does thanks!
